### DISCUSSION TITLE
0016 XLS-16d: NFT Metadata

### DISCUSSION THREAD

**Discussion by HubertG97** [XRPLF/XRPL-Standards#37] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by yyolk** [comment]
I like it! 

My one inclination is thinking that the protocol scheme in the URI can indicate whether or not it's IPFS.

I think adopting IPFS directly is not a bad idea, but it's just a URI with a scheme. 

Maybe I'm the only one that thinks this, but I've been perfectly capable of storing assets for millions of users for years (longer) and for much cheaper than IPFS.

In that regard, my suggestion would be to recommend something like IPFS for "primary" URI, but not stipulate it.

```NFT/2 - Primary URI - text/uri```

**Reply by yyolk** [reply]
Not to distract from this proposal, but I understand IPFS benefits for NFT as thus:

- Immutability (the object never changes once it's uploaded, (it's just a URI with the file checksum, intrinsicly can't change)
- Dedupe (well if we're storing the same file (since our pointer is a checksum) we'll use the same pointer - this also can happen at a fs-block level)
- Decentralized storage
  - like a torrent, your files become seeded on the interplanetary torrent system, but if nobody looks at them, they'll get de-prioritized and it'll be harder to get it out
  - most expensive way to store something
- [not proven] *the idea of future custodianship* in the future, IPFS will never go away and the data will persist forever (you may have to pay a lot to get it out though)
  - S3 on the other hand gets cheaper all the time and is more deeply embedded than IPFS currently

**Reply by HubertG97** [reply]
Thank you! I agree, people can choose whatever URI they want to use, but Indeed above mentioned points make it very suitable for the usage of NFTs

**Reply by calvincs** [reply]
I like the points detailed here: https://medium.com/pinata/who-is-responsible-for-nft-data-99fb4e8147e4
While local hosting can be cheaper, eventually, your domain will need to be renewed, DNS can go down, etc. It relies on a system that has many points that can break.  That said, having multiple "seed" sources is likely a good idea.  And at some point, someone needs to pay for hosting that data.
Having NFTs that generate revenue that allows them to pay for their own existence could be one way to help ease this problem.  IPFS is making that possibility easier with remote pinning services.  NFTs + ILP viewing -> partially paying for NFT hosting seems viable. 
It might also be worth adding WebTorrents into the mix as a secondary source of NFT data. 
https://webtorrent.io/

**Reply by calvincs** [reply]
Also should mention, webtorrents DO allow failback HTTPS seed sources (like domains), just as normal torrents do.
http://www.bittorrent.org/beps/bep_0019.html
https://github.com/webtorrent/create-torrent

Pretty easy to write up, could add it to my example codebase @ https://github.com/calvincs/xrpl-nft-creator

**Reply by yyolk** [reply]
This is really good discussion thank you for the feedback!

I would feel remiss if I didn't mention that IPFS is directly correlated with FILecoin. Therefore, buying into IPFS will directly lead to buy in of $FIL. I don't wish to support the business model of $FIL which hasn't been proven. 

> While interacting with IPFS does not require using Filecoin, all Filecoin nodes are IPFS nodes under the hood, and can connect to and fetch IPLD-formatted data from other IPFS nodes using libp2p. However, Filecoin nodes don't join or participate in the public IPFS DHT. 

"IPFS Distrubuted Hash Table" = torrent tracker


In lieu of [today's events](https://www.coindesk.com/filecoin-double-deposit-on-binance-exploit-open-other-exchanges) one can reason why $FIL doesn't directly call out it's intrinsically tied to IPFS.

I'd rather support protocols like torrents like @calvincs suggested, which is what IPFS is wrapping and calling a protocol. 

The metadata store itself or pinning metadata is not necessary when we could include an alternative like magnet-links.

Sorry for diverging but I think defaults often imply trust and best option, when I, who is probably a minority, don't accept that's the case. 

For me referring it to `Primary URI` over `IPFS URI` is thinking forward towards integrations requiring that to be IPFS, if it were to be suggested and accepted here. 

When instead we could interpret like another web resource with a few suggestions for primary asset store, I also am inclined to believe that can change over time, and new things can be supported - with this suggestion being the least amount of friction for those already in NFT (albeit, in my opinion, a bit jaded with what they expect vs what is reality).

**Reply by HubertG97** [reply]
I just read your tweet about it and could create concern's using IPFS for NFTs. I will for sure try out using a torrent protocol. Also the majority of people are unaware of the storage of the actual file or image of a NFT. 
Now we are just starting with NFTs on the XRPL we of course want to do it the best way possible and find the most reliant way of keeping the data available. 
I also agree that these services will come and go, still hope for all the NFTs already minted on the ETH chain that IPFS nodes will stay up, and don't make the NFTs worthless.

**Reply by calvincs** [reply]
I do want to make sure we all understand that IPFS came before Filecoin.  And BitTorrent before Justin S. from Tron bought it and eventually re-invented the stack to incorporate IPFS tech (BTFS).  

While Filecoin in part runs on IPFS, IPFS does not require Filecoin.  Same for BitTorrent.  ETH also uses IPFS, and can use other protocols as well.  

I would want to steer away from not using great technologies just because another technology uses the same protocol. The more we can incorporate Web3.0 tech, the better honestly.  More of an "All ships rise" mentality.

**Reply by yyolk** [reply]
A few things i would be remiss if I didn't remark.

- BitTorrent is a protocol made by Bram Cohen. The company of BitTorrent may have been bought but that doesn't meant he owns the standard in the "protocol."

- in their own FAQ IPFS and FileCoin were made by the same company. 
![6A1FDC8C-C708-4367-B610-B3FA9F954C11](https://user-images.githubusercontent.com/134478/112053204-0e4cdf80-8b22-11eb-9452-0c043de9b4f6.jpeg) via their [FAQ](https://docs.ipfs.io/concepts/faq/#contributing-to-ipfs)

- Web3 is a facade of a standard intentionally made around buzzwords like web 3.0, which is semantic web. don't mistake web3 for a standard as well, it's a poor mish mash of bad user experiences, entirely geared for ETH. XUMM for instance uses recognizable flows, but is all done uniquely better IMO, and is similar to signing in to WhatsApp or discord through a QR code.


Apologies for tossing around words like standard/protocol, it's not very helpful and am trying to arrive at a better idea myself. I also may be a bit hyperbolic but I don't necessarily agree with "let's just copy what ETH does.", it's hard for me to accept that we have something good / emblematic to look at when it comes to ETH's NFT for good user experiences.

**Comment by shortthefomo** [comment]
An important read https://twitter.com/jonty/status/1372163423446917122 some of the ISSUES with NFT's

**Reply by calvincs** [reply]
Agree with all these concerns; let me address my thoughts on a few of them.

In the link above yours, "Who is responsible for NFT data", they bring up a good point; at some time, another person will be responsible for maintaining the property.  
Whether it be a physical copy or digital (which you could argue is still physical in some sense), someone needs to maintain that property.  
 
With that in mind, the property itself shouldn't have to live on the wire to be used or considered valuable.  

Some of my thoughts on this:

1) The owner of the NFT should own the responsibility to maintain that property. That is their responsibility. That needs to be clearly defined in a legal section within the NFTs metadata.

1a) NFT metadata should contain legal verbiage that spells out the responsibilities of all parties.

2) We, as developers of an NFT should provide as many ways to decentralize the initial data storage as possible, within reason. And fully express this limitation to the end-user for awareness (see above 1a) IPFS/WebTorrents/HTTP/others 

3) All NFTs  files should, in their metadata, contain a link back to the "proof" of ownership that exists on a chain, whether it be XRP, ETH, or any other chain.   As an example, see the metadata in the mp4 file https://ipfs.io/ipfs/bafybeiccp7ikixrum3kizkbezqtvl4fylomiihdtg4pm2wyhojhsp3wdgu/BigBuckBunny.mp4
It contains a string "XRP://XVNFT2byUWwpzimqh1LjN2RcrYBNbymJEChnuSrE1n4n8jH" and other details which help prove our ownership.
This should be a standard for NFTs  content in general, IMHO.

4) Optional but undoubtedly beneficial, having a multichain NFT would serve a similar purpose as data storage in proving ownership and building in the resiliency of proof across multiple chains.

5) All NFTs should provide multiple hashes of file content which helps prove the data is owned by the NFT holder, despite the platform. As such CIDs, SHA256 checksums, WebTorrents all provide this in some fashion. 
These hashes should be easily visible in the blockchains they are associated with for easy validation of content ownership.

A downloaded copy of the work will have the same SHA256 as the original and still allows the owner to retain the work offline if they so choose.

6) As developers, we should support tooling which easily allows users to retain local copies of their NFTs.  These local copies should support re-pinning or hosting the content and, or if they so choose, leaving it offline.  But if even offline, having the ability to prove cryptographically they own said works.

**Comment by calvincs** [comment]
Transferring thoughts from another thread: https://github.com/XRPLF/XRPL-Standards/discussions/34#discussioncomment-505560

If the CTI pointer is failing to resolve to the meta, how does the end users application pull the meta that it requires in order to show the content, or validate they own the digital objects? 

This is one thing I really like about the XRP ledger, and I find as one of its strong suites as compared to other block chains, you don't require a full history of TXs. If we begin to use the meta in the memo of a TX, and then use a pointer which can roll off (unless you maintain full history), you create complexity. Over time the system will become fragile (NFTs).

However, if the pointer/meta is in the Account Root, no number of transactions is required for storage for later retrieval.
Are my assumptions of this limitation incorrect?

If not, I think we should really consider putting the meta/pointer data in the Account Root, like the Domain/Email/Msg Key hash, or create a new field (I dont think this is required tho).

Im writing a piece of software which may help with this by providing Account Root summary data in a p2p fashion for clients. And while some hosts would need to periodically query the ledger to compile said data, those hosts would not require the full processing power of a full history rippled node to do that, as we are only examining the latest state of the ledgers accounts. And clients as simple as phone web browsers should be able to acquire, parse, and locate whatever services or meta they require from the ledger, from a given address, in a reasonable time period.

If you can use the strengths the XRP ledger gives you out of the box, and boot strap your apps that way instead of requiring heavy services like full history nodes, IMHO it makes the app and the use cases stronger. Small light weight apps should have as easy of a time locating services or pointers tied to assets on the ledger as possible w/o requiring resource intensive infrastructure.

I understand Xumm has a well maintained back end, full history nodes, beefy post processing w/ lookup tables, etc. But not all apps have that power, and if possible, if we can make that not a dependency for applications, XRP dependent apps and services would be better for it. This would also allow more agility for apps, less dependence on said heavy services like full history nodes do function normally.

This is why when I built the demo code I did, it used the Account Root, the one thing that always carries forward.
Thoughts?

**Reply by RichardAH** [reply]
> Transferring thoughts from another thread: [#34 (reply in thread)](https://github.com/XRPLF/XRPL-Standards/discussions/34#discussioncomment-505560)
> 
> If the CTI pointer is failing to resolve to the meta, how does the end users application pull the meta that it requires in order to show the content, or validate they own the digital objects?
> 
> This is one thing I really like about the XRP ledger, and I find as one of its strong suites as compared to other block chains, you don't require a full history of TXs. If we begin to use the meta in the memo of a TX, and then use a pointer which can roll off (unless you maintain full history), you create complexity. Over time the system will become fragile (NFTs).
> 
> However, if the pointer/meta is in the Account Root, no number of transactions is required for storage for later retrieval.
> Are my assumptions of this limitation incorrect?
> 
> If not, I think we should really consider putting the meta/pointer data in the Account Root, like the Domain/Email/Msg Key hash, or create a new field (I dont think this is required tho).

If your concern is that there will be no full history servers and full history will be lost then you should be equally concerned that there will be no validators and the current state will also be lost.

Using data stored in (current) account state means the data is mutable. It can always be changed after the token is issued by the owner of the account unless the account is blackholed. It is also a very limited storage, much smaller and less versatile than a txn memo.

My suggestion for making CTI based metadata more “permanent” if you are worried about history loss is to record the full canonical ledger for the CTI along with the signed validation of it by the then-current UNL. This can be placed into an NFT certificate in the form of a PDF or similar and given in hard copy if needed as permanent proof of issuance.

**Reply by calvincs** [reply]
I'm not honestly overly concerned about lack of full history nodes, but rather that not all people will have the same access to that data.  And requiring that data, in the long term, adds complexity to the system by requiring that full history nodes would then need to exists in some fashion.  Its even possible requiring full history nodes, due to the cost associated with their storage creates some form of centralization, or in balances due to who has the ability to host such data. I think adding the signed validation by the then-current UNL will be helpful, but still not completely resolve that issue.  

While the account root data is mutable, the account address itself is not.  If the Digital object we create is pointing to the account address, and the account address is pointing back to the Digital object, they now have a relationship that confirms their authenticity.  And while I do understand that the account domain could be changed, I actually see this as a positive thing.  This allows the person with the Digital object to check the account, see the meta link is broken, and therefore unless the domain is set properly, the contract between those two items can be voided.   The owner of the digital object can *release* the Digital Object back to the ecosystem (think public domain) for all users, and reclaim the XRP value from the account deletion.  I think this is a powerful concept, among others this allows.

**Reply by RichardAH** [reply]
> While the account root data is mutable, the account address itself is not. If the Digital object we create is pointing to the account address, and the account address is pointing back to the Digital object, they now have a relationship that confirms their authenticity. And while I do understand that the account domain could be changed, I actually see this as a positive thing. This allows the person with the Digital object to check the account, see the meta link is broken, and therefore unless the domain is set properly, the contract between those two items can be voided. The owner of the digital object can _release_ the Digital Object back to the ecosystem (think public domain) for all users, and reclaim the XRP value from the account deletion. I think this is a powerful concept, among others this allows.

I don't think this is an either/or thing. Taken together XLS14d, XLS15d and XLS16d describe one way to publish NFTs on the XRPL. There's nothing to prevent you from publishing for example an XLS18d as a competing or alternative standard. Provided the first byte of currency code is always a disambiguator then multiple standards can co-exist. It's up to wallets to implement them and authors to use them.

**Comment by intelliot** [comment]
The "blogpost" link goes to https://coil.com/blogs/edit-post/ckljw6y1a9dba0894xk8inqa0 which is not publicly accessible (I get a 404 error).

**Reply by HubertG97** [reply]
Thankyou! I will edit it right now

**Comment by mvadari** [comment]
@HubertG97 should this spec be moved into PR and kept in the repo or should it be closed? Note that this does not mean that the spec has been accepted/implemented/finalized.

I will move it to PR by copy-pasting the existing draft if there is no response in the next week.
